ILSpy.ReadyToRun/ReadyToRunLanguage.cs

```diff
@@ -0,0 +1,170 @@
+					r2rReaders.Add(module, result);
+				}
+			}
+			return result;
+		}
+
+		public override ProjectId DecompileAssembly(LoadedAssembly assembly, ITextOutput output, DecompilationOptions options)
+		{
+			PEFile module = assembly.GetPEFileOrNull();
+			R2RReader reader = GetReader(assembly, module);
+
+			WriteCommentLine(output, "TODO - display ready to run information");
+			// TODO: display other header information
+			foreach (var method in reader.R2RMethods) {
+				WriteCommentLine(output, method.SignatureString);
+			}
+
+			return base.DecompileAssembly(assembly, output, options);
+		}
+
+		public override void DecompileMethod(IMethod method, ITextOutput output, DecompilationOptions options)
+		{
+			PEFile module = method.ParentModule.PEFile;
+			R2RReader reader = GetReader(module.GetLoadedAssembly(), module);
+			int bitness = -1;
+			if (reader.Machine == Machine.Amd64) {
+				bitness = 64;
+			} else if (reader.Machine == Machine.I386) {
+				bitness = 32;
+			}
+			else {
+				// TODO: Architecture other than x86/amd64
+				throw new NotImplementedException("");
+			}
+			foreach (var m in reader.R2RMethods) {
+				if (m.MethodHandle == method.MetadataToken) {
+					// TODO: Indexing
+					foreach (RuntimeFunction runtimeFunction in m.RuntimeFunctions) {
+						WriteCommentLine(output, m.SignatureString);
+						byte[] code = new byte[runtimeFunction.Size];
+						for (int i = 0; i < runtimeFunction.Size; i++) {
+							code[i] = reader.Image[reader.GetOffset(runtimeFunction.StartAddress) + i];
+						}
+						DecoderFormatterExample(output, code, bitness, (ulong)runtimeFunction.StartAddress);
+						output.WriteLine();
+					}
+				}
+			}
+		}
+
+		private void DecoderFormatterExample(ITextOutput output, byte[] exampleCode, int exampleCodeBitness, ulong exampleCodeRIP)
+		{
+			// TODO: Decorate the disassembly with Unwind, GC and debug info
+			// You can also pass in a hex string, eg. "90 91 929394", or you can use your own CodeReader
+			// reading data from a file or memory etc
+			var codeBytes = exampleCode;
+			var codeReader = new ByteArrayCodeReader(codeBytes);
+			var decoder = Decoder.Create(exampleCodeBitness, codeReader);
+			decoder.IP = exampleCodeRIP;
+			ulong endRip = decoder.IP + (uint)codeBytes.Length;
+
+			// This list is faster than List<Instruction> since it uses refs to the Instructions
+			// instead of copying them (each Instruction is 32 bytes in size). It has a ref indexer,
+			// and a ref iterator. Add() uses 'in' (ref readonly).
+			var instructions = new InstructionList();
+			while (decoder.IP < endRip) {
+				// The method allocates an uninitialized element at the end of the list and
+				// returns a reference to it which is initialized by Decode().
+				decoder.Decode(out instructions.AllocUninitializedElement());
+			}
+
+			// Formatters: Masm*, Nasm*, Gas* (AT&T) and Intel* (XED)
+			// TODO: DecompilationOptions?
```

Comment:
You would have to provide your own decompilation options here as most (if not all options) provided by ILSpy won't apply to R2R. You can add a new tab page to the options dialog.

Reply:
Following the TestPlugin implementation, I am able to find a sample on how to do it.
I am a bit concerned about pulling in the WPF dependencies, as it will not work cross-platform.

Being able to customize what disassembly format to use doesn't seem to be a blocking issue, let's address this later?

Reply:
> it will not work cross-platform.

Why would that be necessary?

Reply:
> > it will not work cross-platform.
> 
> Why would that be necessary?

It isn't necessary, but it is desirable.

In the general theme of things, ready to run is a technology that is useful to improve managed processes startup time by avoiding JIT compilation. This is particularly useful for both Windows client applications (e.g. WinForm/WPF) or microservices (e.g. ASP.NET on Docker). The latter is often run on Linux and it would be highly desirable to be able to perform the diagnostics work on a Linux box.

The underlying library was built targetting .NET standard 2.0, which allows it to run cross-platform. I understand the ILSpy in this repo doesn't (it currently target .NET 4.7.2). I am aware of [AvaloniaILSpy](https://github.com/icsharpcode/AvaloniaILSpy) but I have never tried my hand on it yet. I am hoping my work will work there too.

My hesitation is mostly based on the fact that I knew nothing about that project yet.

Reply:
Yes, lots of workloads run on Linux, but usually the (visual) tooling doesn't (devs continue using Windows to build dockerized solutions) or simply isn't used (the adoption numbers aren't exactly great for AvaloniaILSpy). Please stick with Windows-ism for the moment, we can later refactor like we did in the past to get an xplat cmdline et al.

Reply:
Sounds like a plan. I will stick with Windows for now.